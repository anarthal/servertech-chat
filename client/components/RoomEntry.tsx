import NameAvatar from "@/components/NameAvatar";
import { useCallback } from "react";

// A clickable chat room card
export default function RoomEntry({
  id,
  name,
  lastMessage,
  selected,
  onClick,
}: {
  id: string;
  name: string;
  lastMessage: string;
  selected: boolean;
  onClick: (roomId: string) => void;
}) {
  const cardStyle = selected
    ? { backgroundColor: "var(--boost-light-grey)" }
    : {};
  const onClickCallback = useCallback(() => onClick(id), [onClick, id]);

  return (
    <div
      className="flex pt-2 pl-3 pr-3 cursor-pointer"
      onClick={onClickCallback}
    >
      <div className="flex-1 flex min-w-0 p-4 rounded-lg" style={cardStyle}>
        <div className="pr-3 flex flex-col justify-center">
          <NameAvatar name={name} />
        </div>
        <div className="flex-1 flex flex-col min-w-0">
          <p className="m-0 font-bold pb-2" data-testid="room-name">
            {name}
          </p>
          <p className="m-0 overflow-hidden text-ellipsis whitespace-nowrap">
            {lastMessage}
          </p>
        </div>
      </div>
    </div>
  );
}
