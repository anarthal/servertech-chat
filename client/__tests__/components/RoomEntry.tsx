import { render, screen } from "@testing-library/react";
import RoomEntry from "@/components/RoomEntry";
import "@testing-library/jest-dom";

describe("RoomEntry", () => {
  test("RoomEntry not selected", () => {
    const onClick = jest.fn();

    // Render
    const { asFragment } = render(
      <RoomEntry
        id="myRoomId"
        name="My Room"
        lastMessage="What happened?"
        onClick={onClick}
        selected={false}
      />,
    );

    // Sanity check
    expect(screen.getByText("My Room")).toBeInTheDocument();
    expect(screen.getByText("What happened?")).toBeInTheDocument();

    // Clicking calls the callback
    screen.getByText("My Room").click();
    expect(onClick).toHaveBeenCalledWith("myRoomId");

    // Snapshot test
    expect(asFragment()).toMatchSnapshot();
  });

  test("RoomEntry selected", () => {
    // Render
    const { asFragment } = render(
      <RoomEntry
        id="myRoomId"
        name="My Room"
        lastMessage="What happened?"
        onClick={jest.fn()}
        selected={true}
      />,
    );

    // Sanity check
    expect(screen.getByText("My Room")).toBeInTheDocument();
    expect(screen.getByText("What happened?")).toBeInTheDocument();

    // Snapshot test
    expect(asFragment()).toMatchSnapshot();
  });
});
