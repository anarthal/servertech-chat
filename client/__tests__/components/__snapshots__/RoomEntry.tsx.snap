// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoomEntry RoomEntry not selected 1`] = `
<DocumentFragment>
  <div
    class="flex pt-2 pl-3 pr-3 cursor-pointer overflow-hidden"
  >
    <div
      class="flex-1 flex p-4 rounded-lg overflow-hidden"
    >
      <div
        class="pr-3 flex flex-col justify-center"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1lyf849-MuiAvatar-root"
        >
          M
        </div>
      </div>
      <div
        class="flex-1 flex flex-col overflow-hidden"
      >
        <p
          class="m-0 font-bold pb-2"
          data-testid="room-name"
        >
          My Room
        </p>
        <p
          class="m-0 overflow-hidden text-ellipsis whitespace-nowrap"
        >
          What happened?
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`RoomEntry RoomEntry selected 1`] = `
<DocumentFragment>
  <div
    class="flex pt-2 pl-3 pr-3 cursor-pointer overflow-hidden"
  >
    <div
      class="flex-1 flex p-4 rounded-lg overflow-hidden"
    >
      <div
        class="pr-3 flex flex-col justify-center"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1lyf849-MuiAvatar-root"
        >
          M
        </div>
      </div>
      <div
        class="flex-1 flex flex-col overflow-hidden"
      >
        <p
          class="m-0 font-bold pb-2"
          data-testid="room-name"
        >
          My Room
        </p>
        <p
          class="m-0 overflow-hidden text-ellipsis whitespace-nowrap"
        >
          What happened?
        </p>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
